lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,307 @@
+                          Type resultType,
+                          SmallVector<ReassociationIndices> reassociation,
+                          Value operand, ForallOp op, IRRewriter &rewriter,
+                          Operation *originalSubviewOp,
+                          int inductionVarIndexBase) {
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(op);
+  auto expandShape = rewriter.create<memref::ExpandShapeOp>(
+      op.getLoc(),
+      MemRefType::get({tensorShape},
+                      dyn_cast<MemRefType>(type).getElementType()),
+      operand, reassociation);
+  expandShape.setStaticOutputShape(tensorShape);
+  rewriter.setInsertionPointToStart(op.getBody());
+  SmallVector<OpFoldResult> strides(tensorShape.size(),
+                                    rewriter.getIndexAttr(1)),
+      sizes, offsets;
+
+  size_t tileSize =
+      tensorShape.size() - dyn_cast<ShapedType>(resultType).getShape().size();
+
+  SmallVector<int64_t> tileSizes;
+  for (size_t i = 0; i < tileSize; i++) {
+    int inductionVarIndex = inductionVarIndexBase * tileSize + i;
+    offsets.push_back(op.getInductionVars()[inductionVarIndex]);
+    sizes.push_back(rewriter.getIndexAttr(1));
+  }
+
+  for (size_t i = tileSize; i < tensorShape.size(); i++) {
+    sizes.push_back(rewriter.getIndexAttr(tensorShape[i]));
+    tileSizes.push_back(tensorShape[i]);
+    offsets.push_back(rewriter.getIndexAttr(0));
+  }
+
+  auto subviewType =
+      MemRefType::get({tileSizes}, dyn_cast<MemRefType>(type).getElementType());
+  auto [originalStride, originalOffset] =
+      getStridesAndOffset(dyn_cast<MemRefType>(subviewType));
+  subviewType = MemRefType::get(
+      {tileSizes}, dyn_cast<MemRefType>(subviewType).getElementType(),
+      StridedLayoutAttr::get(rewriter.getContext(), ShapedType::kDynamic,
+                             originalStride));
+  auto subview = rewriter.create<memref::SubViewOp>(
+      op.getLoc(), dyn_cast<MemRefType>(subviewType), expandShape.getResult(),
+      offsets, sizes, strides);
+  originalSubviewOp->replaceAllUsesWith(subview);
+}
+
+static LogicalResult insertParallelLoop(ForallOp op, unsigned mTileShape,
+                                        unsigned nTileShape) {
+  xsmm::BrgemmOp brgemmOp = NULL;
+  OpBuilder b(op);
+  IRRewriter rewriter(b.getContext());
+  for (auto oper = op.getBody()->getOperations().begin();
+       oper != op.getBody()->getOperations().end(); oper++)
+    if (dyn_cast<xsmm::BrgemmOp>(oper)) {
+      brgemmOp = dyn_cast<xsmm::BrgemmOp>(oper);
+      break;
+    }
+  if (brgemmOp == NULL)
+    return rewriter.notifyMatchFailure(op, "require brgemm op");
+
+  if (mTileShape == 0 || nTileShape == 0)
+    return rewriter.notifyMatchFailure(op, "require tile shape to not be zero");
+
+  ArrayRef<int64_t> mShape;
+  auto mDefiningOp = brgemmOp.getOperand(1).getDefiningOp();
+  if (isa<memref::SubViewOp>(mDefiningOp))
```

Comment:
This repetition calls for automation. Either a static function or a lambda would be fine.